toolchain/build-toolchain.sh

```diff
@@ -30,76 +45,152 @@ GCC=gcc-9.3.0
 BUILD_ROOT="$2"
 TOOLCHAIN_PREFIX="${BUILD_ROOT}/${TARGET}"
-MAKEOPTS="-j9 -s"
+MAKEFLAGS="-j9 -s"
+export MAKEFLAGS
 
-log "downloading packages"
 mkdir -p "${TOOLCHAIN_PREFIX}"
 cp ./*.patch "${BUILD_ROOT}"
 cd "${BUILD_ROOT}"
 
-[ ! -f ${BINUTILS}.tar.bz2 ] && wget "http://ftp.gnu.org/gnu/binutils/${BINUTILS}.tar.bz2"
-[ ! -f ${GCC}.tar.xz ] && wget "http://www.mirrorservice.org/sites/ftp.gnu.org/gnu/gcc/${GCC}/${GCC}.tar.xz"
+download_and_patch() {
+    log "downloading packages"
 
-log "extracting packages"
-[ ! -d ${BINUTILS} ] && tar jxf ${BINUTILS}.tar.bz2
-[ ! -d ${GCC} ] && tar Jxf ${GCC}.tar.xz
+    [ ! -f ${BINUTILS}.tar.bz2 ] && wget "http://ftp.gnu.org/gnu/binutils/${BINUTILS}.tar.bz2"
+    [ ! -f ${GCC}.tar.xz ] && wget "http://www.mirrorservice.org/sites/ftp.gnu.org/gnu/gcc/${GCC}/${GCC}.tar.xz"
 
+    log "extracting packages"
+    [ ! -d ${BINUTILS} ] && tar jxf ${BINUTILS}.tar.bz2
+    [ ! -d ${GCC} ] && tar Jxf ${GCC}.tar.xz
+}
 
-log "patching ${BINUTILS}"
-for patchfile in "${BINUTILS}"-*.patch; do
-    if [ ! -f "${BINUTILS}/$patchfile.applied" ]; then
-        patch -d "${BINUTILS}" -p1 < "$patchfile"
-        touch "${BINUTILS}/$patchfile.applied"
-    fi
-done
+build_binutils() {
+    log "patching ${BINUTILS}"
+    for patchfile in "${BINUTILS}"-*.patch; do
+        if [ ! -f "${BINUTILS}/$patchfile.applied" ]; then
+            patch -d "${BINUTILS}" -p1 < "$patchfile"
+            touch "${BINUTILS}/$patchfile.applied"
+        fi
+    done
+
+    log "building binutils"
+    rm -rf "${BINUTILS}/build"
+    mkdir -p "${BINUTILS}/build"
+    pushd "${BINUTILS}/build" > /dev/null
+
+    ../configure --target="${TARGET}" --prefix="${TOOLCHAIN_PREFIX}" \
+                 --with-sysroot="${TOOLCHAIN_PREFIX}/${TARGET}" --enable-lto
+    make
+
+    log "installing binutils"
+    make install
+    popd > /dev/null
+}
 
-log "building binutils"
-rm -rf "${BINUTILS}/build"
-mkdir -p "${BINUTILS}/build"
-cd "${BINUTILS}/build"
+build_gcc_stage1() {
+    log "downloading GCC dependencies"
+    (cd "$GCC" && ./contrib/download_prerequisites)
+
+
+    log "patching ${GCC}"
+    for patchfile in "${GCC}"-*.patch; do
+        if [ ! -f "${GCC}/$patchfile.applied" ]; then
+            patch -d "${GCC}" -p1 < "$patchfile"
+            touch "${GCC}/$patchfile.applied"
+        fi
+    done
+
+    log "building GCC (stage1)"
+    rm -rf "${GCC}/build"
+    mkdir -p "${GCC}/build"
+    cd "${GCC}/build"
+
+    # GCC compilation options
+    # --with-sysroot -> cross-compiler sysroot
+    # --with-newlib -> do note generate standard library includes by fixincludes, do not include _eprintf in libgcc
+    # --disable-libssp -> stack smashing protector library disabled
+    # --disable-nls -> all compiler messages will be in english
+    # --disable-tls -> disable Thread Local Storage as it's not supported Yet (note: C11 incompatibility?)
+    # --enable-initfini-array -> force init/fini array support instead ofg .init .fini sections
```

Comment:
yeah, I hope it will stop producing obsolete `.init` and `.fini` sections (which are there because of crtX.o generated by gcc). We will see.

There is also issue with TM clones, don't want to dig into that now (for nice approach would probably require another patch, I see arm-none-eabi handles it in some hackish way -> https://github.com/xpack-dev-tools/arm-gcc-original-scripts/blob/master/build-toolchain.sh#L448).